File: backend/internal/services/sse.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

// GuestEvent represents a guest mutation event for SSE broadcasting.
type GuestEvent struct {
	Type      string          `json:"type"` // "create", "update", "delete", "checkin", "checkout", "seat_assign"
	GuestID   string          `json:"guestId"`
	WeddingID string          `json:"weddingId"`
	Guest     *GuestEventData `json:"guest,omitempty"`
	Timestamp int64           `json:"timestamp"`
}

// GuestEventData is the guest payload in an SSE event.
type GuestEventData struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Phone       string   `json:"phone"`
	Email       string   `json:"email"`
	Pax         int      `json:"pax"`
	RSVP        string   `json:"rsvp"`
	IsVip       bool     `json:"isVip"`
	Notes       string   `json:"notes"`
	Dietary     []string `json:"dietary"`
	TableID     *string  `json:"tableId"`
	SeatNum     *int     `json:"seatNum"`
	CheckedInAt *string  `json:"checkedInAt"`
	AngbaoAmt   *int     `json:"angbaoAmt"`
	GiftItem    *string  `json:"giftItem"`
}

// SSEClient represents a connected SSE client.
type SSEClient struct {
	ID        string
	WeddingID uuid.UUID
	Chan      chan []byte
	Done      chan struct{}
	doneOnce  sync.Once
}

// SSEHub manages SSE connections per wedding and Redis pub/sub.
type SSEHub struct {
	mu      sync.RWMutex
	clients map[uuid.UUID]map[string]*SSEClient // weddingID -> clientID -> client
	redis   *redis.Client
	pubsub  *redis.PubSub
	ctx     context.Context
	cancel  context.CancelFunc
	// ponytail: closed after PSubscribe ACKs so the first Publish is not lost during Redis warm-up.
	// Race B (client joins SSE mid-publish) is not addressed here — fix needs a per-wedding replay buffer.
	ready chan struct{}
}

// NewSSEHub creates a new SSEHub and starts the Redis subscriber.
func NewSSEHub(rdb *redis.Client) *SSEHub {
	ctx, cancel := context.WithCancel(context.Background())
	hub := &SSEHub{clients: make(map[uuid.UUID]map[string]*SSEClient), redis: rdb, ctx: ctx, cancel: cancel, ready: make(chan struct{})}
	go hub.subscribeRedis()
	select {
	case <-hub.ready:
	case <-time.After(5 * time.Second):
		log.Printf("SSE: PSubscribe not ready after 5s; Publish will error until Redis recovers")
	}
	return hub
}

// subscribeRedis listens on the wildcard Redis channel and fans out to local clients.
func (h *SSEHub) subscribeRedis() {
	h.pubsub = h.redis.PSubscribe(h.ctx, "wedding:*:guests")
	// Wait for Redis to confirm the subscription is active
	if err := h.pubsub.Ping(h.ctx); err != nil {
		log.Printf("SSE: PSubscribe ping failed: %v", err)
	}
	close(h.ready)
	ch := h.pubsub.Channel()
	for {
		select {
		case <-h.ctx.Done():
			return
		case msg, ok := <-ch:
			if !ok {
				return
			}
			wid := extractWeddingID(msg.Channel)
			if wid == uuid.Nil {
				continue
			}
			h.mu.RLock()
			clients := make([]*SSEClient, 0, len(h.clients[wid]))
			for _, client := range h.clients[wid] {
				clients = append(clients, client)
			}
			h.mu.RUnlock()
			for _, client := range clients {
				select {
				case client.Chan <- []byte(msg.Payload):
				default:
					// A dropped event leaves the snapshot stale; force reconnect/resync instead.
					h.Unsubscribe(client)
				}
			}
		}
	}
}

// Subscribe registers a new SSE client for a wedding.
func (h *SSEHub) Subscribe(weddingID uuid.UUID) *SSEClient {
	client := &SSEClient{ID: uuid.New().String(), WeddingID: weddingID, Chan: make(chan []byte, 64), Done: make(chan struct{})}
	h.mu.Lock()
	if h.clients[weddingID] == nil {
		h.clients[weddingID] = make(map[string]*SSEClient)
	}
	h.clients[weddingID][client.ID] = client
	h.mu.Unlock()
	log.Printf("SSE: client %s connected for wedding %s", client.ID, weddingID)
	return client
}

// Unsubscribe removes an SSE client.
func (h *SSEHub) Unsubscribe(client *SSEClient) {
	h.mu.Lock()
	if clients, ok := h.clients[client.WeddingID]; ok {
		delete(clients, client.ID)
		if len(clients) == 0 {
			delete(h.clients, client.WeddingID)
		}
	}
	h.mu.Unlock()
	client.doneOnce.Do(func() { close(client.Done) })
	log.Printf("SSE: client %s disconnected from wedding %s", client.ID, client.WeddingID)
}

// Publish sends a guest event to Redis for cross-instance broadcasting.
func (h *SSEHub) Publish(ctx context.Context, weddingID uuid.UUID, event GuestEvent) error {
	select {
	case <-h.ready:
	default:
		return fmt.Errorf("SSE hub not ready")
	}
	data, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("marshal guest event: %w", err)
	}
	return h.redis.Publish(ctx, fmt.Sprintf("wedding:%s:guests", weddingID.String()), data).Err()
}

// Shutdown gracefully stops the Redis subscriber.
func (h *SSEHub) Shutdown() {
	h.cancel()
	if h.pubsub != nil {
		h.pubsub.Close()
	}
}

func extractWeddingID(channel string) uuid.UUID {
	parts := strings.Split(channel, ":")
	if len(parts) != 3 {
		return uuid.Nil
	}
	wid, err := uuid.Parse(parts[1])
	if err != nil {
		return uuid.Nil
	}
	return wid
}

File: frontend/src/lib/api/sse.ts
import { getAuth } from '$lib/stores';

export interface GuestEvent {
	type: 'create' | 'update' | 'delete' | 'checkin' | 'checkout' | 'seat_assign';
	guestId: string;
	weddingId: string;
	guest?: {
		id: string;
		name: string;
		phone: string;
		email: string;
		pax: number;
		rsvp: string;
		isVip: boolean;
		notes: string;
		dietary: string[];
		tableId: string | null;
		seatNum: number | null;
		checkedInAt: string | null;
		angbaoAmt: number | null;
		giftItem: string | null;
	};
	timestamp: number;
}

type EventHandler = (event: GuestEvent) => void;
type StatusHandler = (status: 'connected' | 'reconnecting', reconnect: boolean) => void;

class SSEClient {
	private eventSource: EventSource | null = null;
	private handlers: Set<EventHandler> = new Set();
	private statusHandlers: Set<StatusHandler> = new Set();
	private reconnectTimeout: ReturnType<typeof setTimeout> | null = null;
	private reconnectDelay = 1000;
	private maxReconnectDelay = 30000;
	private wid = '';

	connect(weddingId: string) {
		if (this.eventSource) this.disconnect();
		this.wid = weddingId;
		const { accessToken } = getAuth();
		if (!accessToken) {
			console.warn('SSE: No access token available');
			return;
		}
		const url = `/api/weddings/${weddingId}/events?token=${encodeURIComponent(accessToken)}`;
		this.eventSource = new EventSource(url);
		this.eventSource.onopen = () => {
			const reconnect = this.reconnectDelay > 1000;
			this.reconnectDelay = 1000;
			this.statusHandlers.forEach((handler) => handler('connected', reconnect));
		};
		this.eventSource.onmessage = (event) => {
			try {
				const data: GuestEvent = JSON.parse(event.data);
				this.handlers.forEach((handler) => handler(data));
			} catch (e) {
				console.error('SSE: Failed to parse event', e);
			}
		};
		this.eventSource.onerror = () => {
			this.statusHandlers.forEach((handler) => handler('reconnecting', false));
			this.scheduleReconnect();
		};
	}

	disconnect() {
		this.eventSource?.close();
		this.eventSource = null;
		if (this.reconnectTimeout) clearTimeout(this.reconnectTimeout);
		this.reconnectTimeout = null;
		this.reconnectDelay = 1000;
	}

	onEvent(handler: EventHandler): () => void {
		this.handlers.add(handler);
		return () => this.handlers.delete(handler);
	}

	onStatus(handler: StatusHandler): () => void {
		this.statusHandlers.add(handler);
		return () => this.statusHandlers.delete(handler);
	}

	private scheduleReconnect() {
		if (this.reconnectTimeout) clearTimeout(this.reconnectTimeout);
		this.reconnectTimeout = setTimeout(() => {
			this.reconnectTimeout = null;
			this.reconnectDelay = Math.min(this.reconnectDelay * 2, this.maxReconnectDelay);
			this.connect(this.wid);
		}, this.reconnectDelay);
	}
}

let sseInstance: SSEClient | null = null;

export function getSSIClient(): SSEClient {
	if (!sseInstance) sseInstance = new SSEClient();
	return sseInstance;
}

export function connectSSE(weddingId: string): SSEClient {
	const client = getSSIClient();
	client.connect(weddingId);
	return client;
}

export function disconnectSSE() {
	sseInstance?.disconnect();
	sseInstance = null;
}
